Azure merge2

milismsft opened the pull request.

milismsft committed:
Refactor azure kato oort and mort into clouddriver-azure
Merge azure-lb payload with the latest in Microsoft/clouddriver master branch
Merge azure-template-refactor payload

milismsft committed:
refactoring of existing code

milismsft committed:
updates

Commit:
Clean up create virtual network creation code. Generalized Create resource from template method.
clean up naming of resources

milismsft committed:
Various fixes

Add support for retrieving the dns name for an Azure load balancer
When we delete an Azure load balancer we also need to delete the public Ip address associated with it

milismsft committed:
Add support for retrieving the tags associated with an Azure load balancer

Commit:
Merge branch 'dns_rename' into azure-merge2

# Conflicts:
#	clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/templates/AzureLoadBalancerResourceTemplate.groovy
#	clouddriver-azure/src/test/groovy/com/netflix/spinnaker/clouddriver/azure/resources/loadbalancer/deploy/templates/AzureLoadBalancerResourceTemplateSpec.groovy

Commit:
Merge branch 'azure-merge2' of https://github.com/Microsoft/clouddriver into azure-merge2

# Conflicts:
#	clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureResourceManagerClient.groovy
#	clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/common/AzureUtilities.groovy

milismsft committed:
Merge with latest from dns_rename branch

Clean up create virtual network creation code. Generalized Create resource from template method.
clean up naming of "peripheral" resources for load balancer (Public IP, DNS name, Front End, etc.)

JargoonPard commented on clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy:
```diff
@@ -18,10 +18,17 @@ package com.netflix.spinnaker.clouddriver.azure.client
+import com.netflix.spinnaker.clouddriver.azure.resources.network.model.AzureVirtualNetworkDescription
+import com.netflix.spinnaker.clouddriver.azure.resources.subnet.model.AzureSubnetDescription
 import com.netflix.spinnaker.clouddriver.azure.security.AzureCredentials
+import groovy.json.JsonBuilder
```
I thought that we had agreed to use Jackson instead.

milismsft replied:
nope, we've always used JsonBuilder (same for aws code). Google is using their own.

JargoonPard replied:
In the committed code there is a mix of jackson and Jsonbuilder.

scotmoor replied:
Jackson == ObjectMapper which is used in a lot of places, including the Azure SDK for Java.

JargoonPard commented on clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy:
```diff
@@ -30,38 +37,213 @@ class AzureNetworkClient extends AzureBaseClient {
+    def result = new ArrayList<AzureSubnetDescription>()
+
+    for (VirtualNetwork item : list) {
+      for (com.microsoft.azure.management.network.models.Subnet itemSubnet : item.subnets) {
```
Do you need the fully qualified name here?

milismsft replied:
Yes. There's a Subnet class that spinnaker defines as well; I want to distinguish between the two.

JargoonPard replied:
But is it used anywhere in this class? Disambiguation should be covered by the import statements.

scotmoor replied:
If you add the import, is it throwing some kind of exception about not being able to resolve which class (spinnaker v. azure) you're referencing? I don't see an import for the spinnaker subnet class so it shouldn't have a conflict

JargoonPard commented on clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy:
```diff
@@ -30,38 +37,213 @@ class AzureNetworkClient extends AzureBaseClient {
+      vnetItem.resourceGuid = item.resourceGuid
+
+      def resultSubnet = new ArrayList<AzureSubnetDescription>()
+      for (com.microsoft.azure.management.network.models.Subnet itemSubnet : item.subnets) {
```
Do you need the fully qualified name here?

milismsft replied:
yes (see the previous comment)

JargoonPard commented on clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy:
```diff
@@ -30,38 +37,213 @@ class AzureNetworkClient extends AzureBaseClient {
+   * @return the dns name of the given load balancer
+   */
+  String getDnsNameForLoadBalancer(AzureCredentials creds, String resourceGroupName, String loadBalancerName) {
+    def loadBalancer = this.getNetworkResourceProviderClient(creds).getLoadBalancersOperations().get(resourceGroupName, loadBalancerName).getLoadBalancer()
```
This looks like duplicate code to line 99 above. Extract to a single method.

milismsft replied:
sounds good.

JargoonPard commented on clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureResourceManagerClient.groovy:
```diff
@@ -31,53 +34,39 @@ import com.microsoft.azure.management.resources.models.ResourceGroupListParamete
+import com.netflix.spinnaker.clouddriver.azure.common.AzureUtilities
 import com.netflix.spinnaker.clouddriver.azure.security.AzureCredentials
 import groovy.transform.Canonical
+import groovy.json.JsonBuilder
```
Switch to Jackson

JargoonPard commented on clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureResourceManagerClient.groovy:
```diff
@@ -87,10 +76,11 @@ class AzureResourceManagerClient extends AzureBaseClient {
+  ResourceGroup createResourceGroup(AzureCredentials creds, String resourceGroupName, String region) {
     try {
-      return this.getResourceManagementClient(creds).getResourceGroupsOperations().createOrUpdate(resourceGroupName, rg).resourceGroup
+      ResourceGroup resourceGroup = this.getResourceManagementClient(creds).getResourceGroupsOperations().createOrUpdate(resourceGroupName,new ResourceGroup(region)).resourceGroup
```
Why assign to a new object and then return that object. Why not just return on this line?

scotmoor replied:
in fact, based on feedback from Netflix in previous PR reviews, shouldn't we also remove the "return" statement?

milismsft replied:
because the return is in the try statement, I think it's fine to leave it there.

JargoonPard commented on clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureResourceManagerClient.groovy:
```diff
@@ -100,7 +90,8 @@ class AzureResourceManagerClient extends AzureBaseClient {
     ResourceGroupListParameters parameters = new ResourceGroupListParameters()
```
Are lines 90-94 reachable?

milismsft replied:
getResourcesGroupsForApp()? not at the moment but I will need that for the catching of server groups/clusters

JargoonPard replied:
Remove unreachable code.

scotmoor replied:
We should remove the getResourcesGroupsForApp since there is a 1:1 correlation now between an app and a resource group. This method wont find you the server groups or clusters.

milismsft replied:
noted for now to be removed from the netflix PR.

JargoonPard commented on clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureResourceManagerClient.groovy:
```diff
@@ -100,7 +90,8 @@ class AzureResourceManagerClient extends AzureBaseClient {
+    this.getResourceManagementClient(creds).getResourceGroupsOperations().list(parameters).resourceGroups
   }
 
   boolean resourceGroupExists(AzureCredentials creds, String resourceGroupName) {
```
This method doesn't make sense to me.  The resourceGroupName isn't used and the return value of the method probably isn't a boolean.

milismsft replied:
see line 61; we create a new resource group the first time we deploy a template if one does not exist already for the given application name and region

JargoonPard replied:
but you are still not using the resourceGroupName parameter and I don't expect the return types match correctly.

JargoonPard commented on clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/common/cache/provider/AzureInfrastructureProvider.groovy:
```diff
@@ -0,0 +1,61 @@
+
+  @Override
+  String getProviderName() {
+    return PROVIDER_NAME
```
Don't need return keyword

JargoonPard commented on clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/common/cache/provider/AzureInfrastructureProvider.groovy:
```diff
@@ -0,0 +1,61 @@
+
+  @Override
+  Map<String, String> parseKey(String key) {
+    return Keys.parse(azureCloudProvider, key)
```
Don't need return keyword

JargoonPard commented on clouddriver-azure/src/test/groovy/com/netflix/spinnaker/clouddriver/azure/resources/loadbalancer/deploy/converters/UpsertAzureLoadBalancerAtomicOperationConverterUnitSpec.groovy:
```diff
@@ -0,0 +1,139 @@
+
+  def setupSpec() {
+    this.converter = new UpsertAzureLoadBalancerAtomicOperationConverter(objectMapper: mapper)
+    def accountCredentialsProvider = Mock(AccountCredentialsProvider)
```
Be consistent in your naming. Add "mock" to this variable name

JargoonPard commented on clouddriver-azure/src/test/groovy/com/netflix/spinnaker/clouddriver/azure/resources/loadbalancer/deploy/converters/UpsertAzureLoadBalancerAtomicOperationConverterUnitSpec.groovy:
```diff
@@ -0,0 +1,139 @@
+      ]
+
+    when:
+      def description = converter.convertDescription(input)
```
There isn't any logic to be tested in this method.

JargoonPard commented on clouddriver-azure/src/test/groovy/com/netflix/spinnaker/clouddriver/azure/resources/loadbalancer/deploy/converters/UpsertAzureLoadBalancerAtomicOperationConverterUnitSpec.groovy:
```diff
@@ -0,0 +1,139 @@
+    // TODO add rest of the checks
+
+    when:
+      def operation = converter.convertOperation(input)
```
Likewise there isn't any logic to be tested in this method either.

JargoonPard commented on clouddriver-azure/src/test/groovy/com/netflix/spinnaker/clouddriver/azure/resources/loadbalancer/deploy/ops/DeleteAzureLoadBalancerAtomicOperationUnitSpec.groovy:
```diff
@@ -0,0 +1,137 @@
+import spock.lang.Specification
+
+class DeleteAzureLoadBalancerAtomicOperationUnitSpec extends Specification {
+  private static final LOAD_BALANCER_NAME = "azureapp1-st1-d1"
```
Is the setup the same between this test and the previous test? If so extract the code into a single location.

JargoonPard commented on clouddriver-azure/src/test/groovy/com/netflix/spinnaker/clouddriver/azure/resources/loadbalancer/deploy/ops/DeleteAzureLoadBalancerAtomicOperationUnitSpec.groovy:
```diff
@@ -0,0 +1,137 @@
+      def operation = converter.convertOperation(input)
+
+    then:
+      operation instanceof DeleteAzureLoadBalancerAtomicOperation
```
Does this actually test any logic in the convertDescription method?

milismsft replied:
Yes, it calls the specific implementation for the validator method and the actual converter.

JargoonPard replied:
Sure it calls code but where's the test that the code does the right thing?

JargoonPard commented on clouddriver-azure/src/test/groovy/com/netflix/spinnaker/clouddriver/azure/resources/loadbalancer/deploy/ops/UpsertAzureLoadBalancerAtomicOperationUnitSpec.groovy:
```diff
@@ -0,0 +1,145 @@
+import spock.lang.Specification
+
+class UpsertAzureLoadBalancerAtomicOperationUnitSpec extends Specification {
+  private static final LOAD_BALANCER_NAME = "azureapp1-st1-d1"
```
as per previous comment regarding duplicate code and extracting.

JargoonPard commented on clouddriver-azure/src/test/groovy/com/netflix/spinnaker/clouddriver/azure/resources/loadbalancer/deploy/ops/UpsertAzureLoadBalancerAtomicOperationUnitSpec.groovy:
```diff
@@ -0,0 +1,145 @@
+    converter.accountCredentialsProvider = accountCredentialsProvider
+  }
+
+  void "create a load balancer dynamic template with two probes and two listners/load balancer rules"() {
```
Your tests don't reflect what you state you are testing.

JargoonPard commented on clouddriver-azure/src/test/groovy/com/netflix/spinnaker/clouddriver/azure/resources/loadbalancer/deploy/validators/UpsertAzureLoadBalancerDescriptionValidatorSpec.groovy:
```diff
@@ -0,0 +1,156 @@
+import spock.lang.Specification
+
+class UpsertAzureLoadBalancerDescriptionValidatorSpec extends Specification {
+  private static final LOAD_BALANCER_NAME = "azureapp1-st1-d1"
```
See previous comment on code duplication and reuse

JargoonPard commented on clouddriver-azure/src/test/groovy/com/netflix/spinnaker/clouddriver/azure/resources/loadbalancer/deploy/validators/UpsertAzureLoadBalancerDescriptionValidatorSpec.groovy:
```diff
@@ -0,0 +1,156 @@
+    converter.accountCredentialsProvider = credentialsProvider
+  }
+
+  void "pass validation with proper description inputs"() {
```
A better test is to ensure the validator correctly detects error conditions.

milismsft replied:
Yes, I agree. I was hopping that Haishi's site will take care of that rather then us duplicating that work. For now since we check against a constant set of values anyway I'm more concern with the execution flow (i.e. calling our Azure specific implementation) than the final content of the description object being returned.

scotmoor commented on clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureResourceManagerClient.groovy:
```diff
@@ -138,7 +133,13 @@ class AzureResourceManagerClient extends AzureBaseClient {
+  private static void createResourceGroupVNet(AzureCredentials creds, String resourceGroupName, String region) {
+    def vNetName = AzureUtilities.VNET_NAME_PREFIX + resourceGroupName
+
+    creds.getNetworkClient().createVirtualNetwork(creds, resourceGroupName, vNetName, resourceGroupName, "10.0.0.0/16")
```
you're passing the resourceGroupName for the region (which is what is causing the createLoadBalancer operation to fail). Also you do not need to address in this call as the method will default to that value.

scotmoor replied:
By the way, I've fixed the call createVirtualNetwork and will push it into azure-merge2. The PR should automatically pick it up

scotmoor commented on clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/loadbalancer/ops/UpsertAzureLoadBalancerAtomicOperation.groovy:
```diff
@@ -51,25 +50,40 @@ class UpsertAzureLoadBalancerAtomicOperation implements AtomicOperation<Map> {
+
+      task.updateStatus(BASE_PHASE, "Beginning load balancer deployment")
+
+      String resourceGroupName = AzureUtilities.getResourceGroupName(description.appName, description.region)
```
The generation of the resource group name should be done in CreateResourceFromTemplate method in AzureResourceManagerClient, otherwise you are going to have to duplicate this code before every call to this method.

scotmoor replied:
I see that further down, we have to pass the resource group name to subsequent calls in this method and unfortunately, there isn't a mechanism to return the resource group name from the call to CreateResourceFromTemplate. For this PR I wouldn't make the change I suggested above but this is definitely something that needs to be addressed as we move forward with creating other resources.

milismsft replied:
we will construct the name of the resource group based on the given description as needed (see AzureUtilities static methods)

milismsft replied:
BTW when we delete a resource, there's no template to be deployed so for consistency purposes I think it's ok to live with that "duplication"

scotmoor commented on kato/kato-azure/kato-azure.gradle:
```diff
@@ -1,7 +0,0 @@
-dependencies {
-  def azureSDKVersion = '0.8.2'
```
Where did this come from? It's not even close to the right version number. Where is it being used?

milismsft replied:
it's not; this stuff got deleted

scotmoor commented on clouddriver-web/config/clouddriver.yml:
```diff
@@ -64,7 +65,7 @@ azure:
 
   accounts:
-    - name: default
+    - name: azure-cred1
```
why are we changing the account name? Should we even be checking in the .yml?

milismsft replied:
this will not go into the netflix PR but since it is checked in as part of my branch changes, it does show in this PR.

Commit:
Fix bug in call to create Virtual Network.
Resolved Merge Conflic in AzureLoadBalancerResourceTemplate

milismsft committed:
fix for vnet creation

milismsft commented on clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy:
```diff
@@ -30,38 +38,214 @@ class AzureNetworkClient extends AzureBaseClient {
+    }
+
+    def publicIpAddressName = AzureUtilities.getResourceNameFromID(loadBalancer.frontendIpConfigurations.first().getPublicIpAddress().id)
+    this.getNetworkResourceProviderClient(creds).getPublicIpAddressesOperations().delete(resourceGroupName, publicIpAddressName)
```
Line 107 and 109 need to be switched, otherwise deletion does not fully work.

milismsft committed:
fix

milismsft committed:
Merge branch 'azure-merge2' of github.com:Microsoft/clouddriver into azure-merge2

Conflicts:
	clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureResourceManagerClient.groovy

milismsft committed:
various fixes and features

Fix the bugs introduced by the previous change (bad merge).
Add Azure Network Security Group upsert and delete related support (no deployment template)

milismsft committed:
Add support for caching and viewing of Azure cloud resources
Small tweaks and bug fixes.
